secure_api/merkle_tree/MerkleTree.Spec.fst

```diff
@@ -45,7 +44,7 @@ let hash2_raw src1 src2 =
 /// requires extending the definitions of tags and hashes. Our
 /// extended definition of hash justifies skipping any concrete
 /// computation on dummies.
-
+noeq
```

Comment:
Because this is client code of `EverCrypt.Hash` it's fine to break the `Lib.IntTypes` abstraction so that this is an `eqtype`, avoiding the use of `prop` and `StrongExcludedMiddle`.

Reply:
I suggested to Aymeric not trying to break the abstraction too much; since we're trying to settle on Lib.IntTypes everywhere, the fewer `friend`s we have, the better -- also, it would require adding an interface to this file

What solution would you envision? Something like Vale.Arch.BufferFriend to centralize the uses of friend? Or just friend in this module?

@R1kM needs to tell us whether the complexity of StrongExcludedMiddle is manageable or not

Reply:
In the current status, I managed to successfully use Lib.IntTypes almost everywhere, and I think StrongExcludedMiddle is ok (it's only for a ghost security proof). The only problem left is MerkleTree.New.Low: the file is pretty big, and currently has no interface, which prevents us from using friend.
As an experiment, I tried to reveal IntTypes (through an assumed squash type equality), but verification failed line 680 in create_empty_mt, and I'm not completely sure why yet, it's probably fixable. 
If somebody knows exactly what needs to be exposed for clients to use Merkle Trees, a solution might be to have a small MerkleTree.New.Low.fsti, and have most of the existing code in a fst with friend Lib.IntTypes

Reply:
There are three places in `MerkleTree` where the secret integers abstraction in `Spec.Hash.Definitions.bytes_hash` needs to be broken:
1. When serializing hashes
2. When deserializing hashes
3. When comparing the computed root of a tree against a purported root

I ported `MerkleTree.New.Low` to use `Lib.IntTypes.uint8` for hashes, breaking the abstraction in the first two cases and declassifying the result of the comparison in the third:
1. This requires calling `Lib.RawIntTypes.u8_to_UInt8` to declassify individual bytes of hashes. The extracted C code remains almost the same (I just had to introduce a temporary variable).
2. Rather than calling `Lib.RawIntTypes.u8_from_UInt8` to endorse individual bytes of hashes, which could have a significant runtime cost if done with a loop, I wrote a separate `Lib.RawBuffer` module that befriends `Lib.IntTypes` and does this calling `LowStar.Buffer.blit` with no change in the generated C code.
3. Since comparing hashes in constant-time is cheap enough, I used `Lib.ByteBuffer.lbytes_eq` to do the comparison in secret hashes, declassifying the result.

This first two changes are completely transparent to clients. The third could have some minor impact, but I don't think this matters for e.g. CCF. 
As an alternative, I implemented a non constant-time comparison in `Lib.RawBuffer.lbytes_eq` that we could use instead to generate exactly the same C as before. To switch to it, uncomment [this line](https://github.com/project-everest/hacl-star/blob/414118ebc486949fb9e08b30077afa5e57159d00/secure_api/merkle_tree/MerkleTree.New.Low.fst#L3054) and comment out the one above.